runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -639,6 +640,164 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+            bool removeCall = false;
+            if (mhConstraint
+                && mhConstraint->getKnownObject()
+                && mhConstraint->isNonNullObject()
+                && mhConstraint->getClass()
+                && comp()->fej9()->isInstanceOf(mhConstraint->getClass(), primitiveHandleClass, true, true) == TR_yes)
+               {
+               uint32_t defcOffset = J9VMJAVALANGINVOKEPRIMITIVEHANDLE_DEFC_OFFSET(comp()->fej9()->vmThread());
+               uintptrj_t* mhLocation = getObjectLocationFromConstraint(mhConstraint);
+
+               TR::VMAccessCriticalSection initializeClassIfRequired(comp(),
+                                                        TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+               if (!initializeClassIfRequired.hasVMAccess())
+                  break;
+               uintptrj_t mhObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mhLocation);
+               uintptrj_t defc = comp()->fej9()->getReferenceFieldAtAddress(mhObject + defcOffset);
+               J9Class* defcClazz = (J9Class*)TR::Compiler->cls.classFromJavaLangClass(comp(), defc);
+               if (defcClazz->initializeStatus == J9ClassInitSucceeded)
+                  {
+                  removeCall = true;
+                  if (trace())
+                     {
+                     traceMsg(comp(), "defc in MethodHandle %p has been initialized\n", mh);
+                     }
+                  }
+               }
+
+            if (removeCall
+                && performTransformation(comp(), "%sRemoving call node %s [" POINTER_PRINTF_FORMAT "]\n", OPT_DETAILS, node->getOpCode().getName(), node))
+               {
+               TR::Node* receiverChild = node->getArgument(0);
+               TR::TransformUtil::transformCallNodeToPassThrough(this, node, _curTree, receiverChild);
+               TR::DebugCounter::incStaticDebugCounter(comp(), TR::DebugCounter::debugCounterName(comp(), "constrainCall/(%s)", signature));
+               return;
+               }
+            break;
+            }
+         case TR::java_lang_invoke_DirectHandle_nullCheckIfRequired:
+            {
+            TR::Node* mh = node->getArgument(0);
+            TR::Node* receiver = node->getArgument(1);
+            bool mhConstraintGlobal, receiverConstraintGlobal;
+            TR::VPConstraint* mhConstraint = getConstraint(mh, mhConstraintGlobal);
+            TR::VPConstraint* receiverConstraint = getConstraint(receiver, receiverConstraintGlobal);
+
+            J9VMThread* vmThread = comp()->fej9()->vmThread();
+            TR_OpaqueClassBlock* directHandleClass = calledMethod->classOfMethod();
+            if (!mhConstraint
+                || !mhConstraint->isNonNullObject()
+                || !mhConstraint->getClass()
+                || comp()->fej9()->isInstanceOf(mhConstraint->getClass(), directHandleClass, true, true) != TR_yes)
+               break;
```

Comment:
This always requires a constraint on `mh` to guarantee it's a (non-null) `DirectHandle`, even when `receiverConstraint->isNonNullObject()` suffices to eliminate the call without reading `final_modifiers`. Was that intentional?

I think it's still possible to remove the call even if `mh` could be null. The call would be under a `NULLCHK`, and replacing it with `PassThrough` should leave that `NULLCHK` in place, now checking `mh` without calling.